### ISSUE TITLE
Variation product negative quantity bug

### ISSUE THREAD

**Issue by pukys** [PrestaShop/PrestaShop#14913]
#### Describe the bug

I have 2 variations at the same product
The language of the employee must be `Lietuvių kalba (Lithuanian)`

```
A variation, quantity = -10
B variation, quantity = 45
```

When I am changing B variation quantity from 45 to 50, prestashop reset A variation quantity from -10 to 0. A variation quantity should remain untouched.

#### Steps to Reproduce

1. Create new product
2. Add 2 new variations
    A variation with quantity -10
    B variation with quantity 10
3. Save product and reload page
4. Edit product again
5. Change variation B quantity from 10 to 15 and click save
6. Reload page (A will have quantity 0, B will have quantity 15).

**Screenshots**

https://drive.google.com/file/d/1Vu6ZKkMo5ijDdDASQ888Oy0F8KTGVJ4V/view

#### Additional information

 * PrestaShop version: 1.7.7.5
 * PHP version: 7.2

**Comment by khouloudbelguith** [comment]
Hi @pukys,

Thanks for your report.
This behavior is reproduced from the BO or the FO?
Could you please provide us with more info? We need more details to understand how we can reproduce your issue.

Thanks!

**Comment by pukys** [comment]
Bug appears in Back Office

1. Create new product
2. Add 2 new variations
    A variation with quantity -10
    B variation with quantity 10
3. Save product and reload page
4. Edit product again
5. Change variation B quantity from 10 to 15 and click save
6. Reload page (A will have quantity 0, B will have quantity 15).

As a result variation A will be quantity 0 changed from -10, though I was editing B variation only.
I have made video how to reproduce it: https://gofile.io/?c=NCjUiO

**Comment by khouloudbelguith** [comment]
@pukys, thanks for your feedback.
I did not manage to reproduce the issue with PS1.7.6.0.
I attached a screen record
https://drive.google.com/file/d/1L1NKMluHIa16e8sdS2Z9qwGOo_KGtl39/view

Could you please provide us with more info:
 - host
 - server setup and configuration
 - PrestaShop version (source)
 - debug mode report
 - PHP error logs
 - apache error log
 - javascript console log
Don't you know how to get this information? Please read the following article:
http://build.prestashop.com/howtos/misc/how-to-create-bug-report/

Thanks!

**Comment by pukys** [comment]
I have installed fresh copy of prestashop on the same hosting, and there everything works fine!
Looks like problem comes from installed modules or store settings. I will keep investigating.

**Comment by khouloudbelguith** [comment]
@pukys, thanks!
Waiting for your feedback.

**Comment by pukys** [comment]
@khouloudbelguith OK, I have found explanation for this bug. It was fault of specific language I am using, switching to English in BO solved the problem.

I was using Lithuanian language, and minus symbol, in localisation files was added with wrong char code. Correct one is 42, in Lithuanian is 8722, and changing it to 42 fixes the problem.

If you would like to reproduce the bug, you should check with Lithuanian lang in BO.

**Comment by khouloudbelguith** [comment]
@pukys, thanks for these clarifications.
I manage to reproduce the issue with PS1.7.6.0 & PS1.7.5.1.
I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!
Thanks!

**Comment by marionf** [comment]
@khouloudbelguith Could you please add a video or describe all the steps to reproduce the issue ?

**Comment by khouloudbelguith** [comment]
@marionf, https://drive.google.com/file/d/1Vu6ZKkMo5ijDdDASQ888Oy0F8KTGVJ4V/view

Thanks!

**Comment by tswfi** [comment]
I can replicate this with Finnish language on version 1.7.7.5 also.

@pukys any pointer on which localisation file the problem lies in?

**Comment by pukys** [comment]
> I can replicate this with Finnish language on version 1.7.7.5 also.
> 
> @pukys any pointer on which localisation file the problem lies in?

Nope, I don't even remember what problem it was, too much time passed :) But as I can see problem was with Lithuanian localisation.

**Comment by tswfi** [comment]
tested on current develop branch and this is a bit nasty.. `-5` converts to `5` when saving the product

edit: and it happens on english language also (default clean git clone && docker-compose up installation)

**Comment by matks** [comment]
@PrestaShop/qa-team Hello, could you please verify the informations from @tswfi 

> edit: and it happens on english language also (default clean git clone && docker-compose up installation)

This could mean a lot more users are impacted than we initially thought

**Comment by hibatallahAouadni** [comment]
Hello @tswfi & @matks 

> tested on current develop branch and this is a bit nasty.. -5 converts to 5 when saving the product
edit: and it happens on english language also

I confirm that :disappointed: I only reproduce it on develop (178x => OK, 177x => OK).
So, this new behavior is a new regression on develop.

IMHO, it will be better if we open a new issue for this new regression, @matks wdyt?

PS: about the original issue, I still reproduce it on 1775 (with Lithuanian language).

**Comment by tswfi** [comment]
@hibatallahAouadni thanks! can you confirm the original issue also on Finnish language on 1775 (and perhaps on other languages too.. if this affects more than two languages there is a systematic error somewhere...)

**Comment by hibatallahAouadni** [comment]
Hello @tswfi 

I reproduce the original issue with **Finnish** and **Arabic** language on 1775, see the attached screen record below:

https://drive.google.com/file/d/1ehQEIRN7mshgiaQrvmVtH5Q7qNEzhb52/view

The issue doesn't occur only with the negative number it happen even with positive number, following these steps:

1. Add 2 new variations
2. A variation with quantity -5
3. B variation with quantity 5
4. Save product and reload page
5. Edit product again
6. Change variation B quantity from -5 to -10 and click save
7. Reload page (A will have quantity -10, B will have quantity 0).

![image](https://user-images.githubusercontent.com/23308427/128676788-33624ea4-ff20-44ee-a6e6-dc46c55f8c36.png)

I didn't reproduce it with **French**, **English**, **Czech** and **Polish**.
@tswfi thanks for reporting the regression on develop, I created an issue #25545 :pray: 

Thanks!

**Comment by tswfi** [comment]
now that the other problem in dev branch has a pr to fix the abs problem (https://github.com/PrestaShop/PrestaShop/pull/25549) I can replicate this language dependent problem in dev branch also.

use language Finnish (or arabic or some other language that has some other character like '−' as minusSign in cldr. Or at least that is my current hypothesis)
```
$ git grep '<minusSign>'|grep -v '-'
core/common/main/et.xml:			<minusSign>−</minusSign>
core/common/main/eu.xml:			<minusSign>−</minusSign>
core/common/main/fa.xml:			<minusSign>‎−</minusSign>
core/common/main/fa.xml:			<minusSign>‎−</minusSign>
core/common/main/fi.xml:			<minusSign>−</minusSign>
core/common/main/fo.xml:			<minusSign>−</minusSign>
core/common/main/gsw.xml:			<minusSign>−</minusSign>
core/common/main/hr.xml:			<minusSign>−</minusSign>
core/common/main/lt.xml:			<minusSign>−</minusSign>
core/common/main/nb.xml:			<minusSign>−</minusSign>
core/common/main/nn.xml:			<minusSign>−</minusSign>
core/common/main/ps.xml:			<minusSign>‎−</minusSign>
core/common/main/rm.xml:			<minusSign>−</minusSign>
core/common/main/se.xml:			<minusSign>−</minusSign>
core/common/main/sl.xml:			<minusSign>−</minusSign>
core/common/main/sv.xml:			<minusSign>؜−</minusSign>
core/common/main/sv.xml:			<minusSign>−</minusSign>
core/common/main/uz_Arab.xml:			<minusSign>‎−</minusSign>
```

The combination listing shows the stock value with normal - sign but the combination edit page uses this '−' which is not handled correctly when saving.

![image](https://user-images.githubusercontent.com/1352132/128718152-b756a5c9-30e3-4a84-b53f-87152f63f90f.png)

![image](https://user-images.githubusercontent.com/1352132/128718183-3811c9ed-cba2-4f43-892d-9540adcfee2b.png)

notice the subtle variation on the minus sign.

On the listing the quantity is rendered with a plain input field but on the combination data page it uses `{{ form_widget(form.attribute_quantity) }}`

Same happens on the new experimental product edit page

![image](https://user-images.githubusercontent.com/1352132/128719069-13e2dd04-b710-4c93-9c51-95707746707c.png)

And this probably affects all number fields that can be negative... Now to just find where the formatting happens...

**Comment by tswfi** [comment]
correction to the previous. In dev branch as the minussign is a different character the saving does not work at all. (the response from the save request is just `[]` without any error information). 

This means that if even one combination has negative stocklevel editing the product is impossible until the negative stock field is fixed by the user.

In expiremental product edit it saving the combination does work (but it does not seem to save combinations stock levels from the combination list).

**Comment by tswfi** [comment]
#25549 against develop branch now fixes this minusSign parsing issue also.

I've heard there probably wont be another 1.7.7.x release so should this issue be closed?

I do have a fix implemented and here is diff against 1.7.7.x branch:

```diff
diff --git a/src/Adapter/Product/AdminProductWrapper.php b/src/Adapter/Product/AdminProductWrapper.php
index 33f695d4c2..37e75f6991 100644
--- a/src/Adapter/Product/AdminProductWrapper.php
+++ b/src/Adapter/Product/AdminProductWrapper.php
@@ -217,7 +217,8 @@ class AdminProductWrapper
         }
 
         if (isset($combinationValues['attribute_quantity'])) {
-            $this->processQuantityUpdate($product, $combinationValues['attribute_quantity'], $id_product_attribute);
+            $quantity = $this->floatParser->fromString($combinationValues['attribute_quantity']);
+            $this->processQuantityUpdate($product, $quantity, $id_product_attribute);
         }
     }
 
diff --git a/src/PrestaShopBundle/Utils/FloatParser.php b/src/PrestaShopBundle/Utils/FloatParser.php
index 0d9402a71e..c6b031436c 100644
--- a/src/PrestaShopBundle/Utils/FloatParser.php
+++ b/src/PrestaShopBundle/Utils/FloatParser.php
@@ -78,6 +78,12 @@ class FloatParser
         // replace arabic numbers by latin
         $value = $this->arabicToLatinNumberConverter->convert($value);
 
+        // CLDR defines few different minusSigns, convert those first
+        // minussigns from CLDR. '−' is the important one, others currently have some js things happening to them
+        $pattern = '/[\-‒⁻₋−➖﹣－]/u';
+        $replacement = '-'; // regular minussign
+        $value = preg_replace($pattern, $replacement, $value);
+
         // remove all non-digit characters
         $split = preg_split('/[^\dE-]+/', $value);

```

**Comment by hibatallahAouadni** [comment]
> I've heard there probably wont be another 1.7.7.x release so should this issue be closed?

Ping @PrestaShop/prestashop-core-developers could someone please answer @tswfi question :pray:

**Comment by matks** [comment]
I'm not sure to fully understand: bugs https://github.com/PrestaShop/PrestaShop/issues/14913 and https://github.com/PrestaShop/PrestaShop/issues/25545 are identical ?

Anyway this is what should be done:
- if there is a bug that can be confirmed on `develop` we must fix `develop` first
- when the bug is fixed on `develop`, we must consider whether 1) it can be backported 2) it should be backported
- it should be backported if it's a regression OR if it's a major/critical bug
- if answers to 1) and 2) is yes, then we'll submit a PR to backport the bugfix to maintained branches

Right now there is no 1.7.7.x released planned because there is no more major/critical bugs known for 1.7.7 . Only minor/trivial bugs so they are planned to be fixed in 1.7.8

**Comment by tswfi** [comment]
They are not identical.

#25545 (using abs in stocklevel which changed negative stock to poitive) "masks" #14913 (minusSign was not correctly used which made the stocklevel zero) in develop branch.

PR #25549 against develop fixes both issues in develop (and the diff above fixes #14913 in 1.7.7.x branch which does not  have the abs() problem)

**Comment by tswfi** [comment]
@hibatallahAouadni about this comment in my pr: https://github.com/PrestaShop/PrestaShop/pull/25549#issuecomment-953946936

I've removed the fixes mention of this issue from the pr https://github.com/PrestaShop/PrestaShop/pull/25549

Now that PR only fixes https://github.com/PrestaShop/PrestaShop/issues/25545

But if we want a fix for this issue I see two options:

1) we close this with "please update to 1.7.8.1" comment if my pr gets into 1.7.8.1
2) another pr with the fix above. BUT which branch should that pr be made against?

Thanks!

**Comment by hibatallahAouadni** [comment]
Hello @tswfi 

> we close this with "please update to 1.7.8.1" comment if my pr gets into 1.7.8.1

Well, your PR target the develop branch, means that your PR will be merged in 8.0.0 :wink: 

Thanks!

**Comment by tswfi** [comment]
> Well, your PR target the develop branch, means that your PR will be merged in 8.0.0 wink

I've been told to always make pr against develop branch and then the stuff that should come in different versions will be "cherry picked" to 1.7.8.x?

**Comment by MatShir** [comment]
Since the new product page is default one and the issue is [partially fixed](https://github.com/PrestaShop/PrestaShop/issues/25545) on the old one on PS 8.1. If the issue appears again on the page, don't hesitate to open the issue again or create a new one :)
